components/suggestion/suggestion_box/index.tsx

```diff
@@ -0,0 +1,240 @@
+// Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
+// See LICENSE.txt for license information.
+import React from 'react';
+import EventEmitter from 'mattermost-redux/utils/event_emitter';
+
+import QuickInput, {QuickInputState} from '../../quick_input';
+
+import {useRefHandler} from './use_ref_handler';
+import {useContainerRef} from './use_container_ref';
+import {useSuggestionHandler} from './use_suggestion_handler';
+import {useInputModifier} from './use_input_modifier';
+import {useInputHandler} from './use_input_handler';
+import {ISuggestionBoxProps, ISuggestionBoxState} from './types';
+
+const SuggestionBox = React.forwardRef((props: ISuggestionBoxProps, ref) => {
+    const qInputRef = React.useRef<HTMLInputElement | any>(null);
+
+    const suggestionReadOut = React.useRef<HTMLDivElement>(null);
+    const [suggestionBoxState, setSuggestionBoxState] = React.useState<ISuggestionBoxState>({
+        focused: false,
+        matchedPretext: [],
+        items: [],
+        terms: [],
+        components: [],
+        selection: '',
+        allowDividers: true,
+        presentationType: 'text',
+    });
+
+    const updateSuggestionBoxState = React.useCallback(
+        /*eslint max-nested-callbacks: [0]*/
+        (newState: Partial<ISuggestionBoxState>) => setSuggestionBoxState((oldState) => ({
+            ...oldState, ...newState,
+        })),
+        [],
+    );
+
+    const [qInputState, setQInputState] = React.useState(QuickInputState.createNew({
+        value: props.value || '',
+    }));
+
+    const qInputRefHandler = useRefHandler(ref, qInputRef);
+    const qInputModifier = useInputModifier(props);
+
+    const pretextRef = React.useRef<string>('');
+    const preventSuggestionListCloseFlag = React.useRef<boolean>(false);
+    const timeoutIdRef = React.useRef<NodeJS.Timeout >();
+
+    const suggestionHandlerRef = useSuggestionHandler(
+        pretextRef, timeoutIdRef, props, suggestionBoxState,
+        updateSuggestionBoxState, setQInputState, qInputModifier, qInputRefHandler,
+    );
+
+    /*eslint max-nested-callbacks: [0]*/
+    React.useEffect(() => {
```

Comment:
Seems like the state setting is convoluted, perhaps replacing with the HTML input will help in this case too?

Reply:
Can you specify here?

Reply:
I don't really have a concrete suggestion but would be nice to reduce the amount of callbacks. Also I was referring to getting rid of QuickInput here as well